### PR TITLE
[SC 4318] Notebook that explains how to use Dataset and Model

### PR DESCRIPTION
**Pull request by AnilSorathiya**
## Internal Notes for Reviewers
* introduction to dataset and model objects notebook

### CODE REVIEW

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 1]
```diff
@@ -0,0 +1,675 @@
...
+    "\n",
+    "**Tests**: A function contained in the ValidMind Developer Framework, designed to run a specific quantitative test on the dataset or model. Tests are the building blocks of ValidMind, used to evaluate and document models and datasets, and can be run individually or as part of a suite defined by your model documentation template.\n",
+    "\n",
+    "**Metrics**: A subset of tests that do not have thresholds. In the context of this notebook, metrics and tests can be thought of as interchangeable concepts.\n",
+    "\n",
+    "**Custom metrics**: Custom metrics are functions that you define to evaluate your model or dataset. These functions can be registered with ValidMind to be used in the platform.\n",
+    "\n",
+    "**Inputs**: Objects to be evaluated and documented in the ValidMind framework. They can be any of the following:\n",
+    "\n",
+    "- **model**: A single model that has been initialized in ValidMind with [`vm.init_model()`](https://docs.validmind.ai/validmind/validmind.html#init_model).\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:67]: Since this notebook is specifically about the model and dataset objects and how they work maybe these should be their own primary bullet points that explains them as their own thing (so the customers dataset and model objects vs the vm_model and vm_dataset objects that are created via vm.init_x().

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.2 Custom test\n",
+    "A custom test is simply a Python function that takes two types of arguments: `inputs` and `params`. The `inputs` are ValidMind objects (`VMDataset`, `VMModel`), and the `params` are additional parameters required for the underlying computation of the test. We will discuss both types of arguments in the following sections.\n",
+    "\n",
+    "Let's start with a custom test that requires only a ValidMind dataset object. In this example, we will check the balance of classes in the target column of the dataset:\n",
+    "\n",
+    "- The custom test below requires a single argument of type `VMDataset` (dataset).\n",
+    "- The `my_custom_metrics.ClassImbalance` is a unique test identifier that can be assigned using the `vm.metric` decorator functionality. This unique test ID will be used in the UI to load test results in the documentation.\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:271]: @cachafla just merged his PR that changes the name from metric to test fyi

```suggestion
    "- The `my_custom_metrics.ClassImbalance` is a unique test identifier that can be assigned using the `vm.test` decorator functionality. This unique test ID will be used in the UI to load test results in the documentation.\n",
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from validmind.vm_models.dataset.dataset import VMDataset\n",
+    "import pandas as pd\n",
+    "\n",
+    "@vm.metric(\"my_custom_metrics.ClassImbalance\")\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:288]:
```suggestion
    "@vm.test(\"my_custom_metrics.ClassImbalance\")\n",
```

same as above

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:353]:
```suggestion
    "@vm.test(\"my_custom_metrics.ClassImbalance\")\n",
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 5]
```diff
@@ -0,0 +1,675 @@
...
+    "print(vm_train_ds)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.5 Custom test using VMDataset and VMModel as inputs\n",
+    "\n",
+    "We will now create a `@vm.metric` wrapper that will allow you to create a reusable metric. Note the following changes in the code below:\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:519]:
```suggestion
    "We will now create a `@vm.test` wrapper that will allow you to create a reusable test. Note the following changes in the code below:\n",
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+    "  - `VMDataset` objects allow you to access the dataset's true (target) values by accessing the `.y` attribute.\n",
+    "  - `VMDataset` objects allow you to access the predictions for a given model by accessing the `.y_pred()` method.\n",
+    "- The function docstring provides a description of what the metric does. This will be displayed along with the result in this notebook as well as in the ValidMind platform.\n",
+    "- The function body calculates the confusion matrix using the `sklearn.metrics.confusion_matrix` function as we just did above.\n",
+    "- The function then returns the `ConfusionMatrixDisplay.figure_` object - this is important as the ValidMind framework expects the output of the custom metric to be a plot or a table.\n",
+    "- The `@vm.metric` decorator is doing the work of creating a wrapper around the function that will allow it to be run by the ValidMind framework. It also registers the metric so it can be found by the ID `my_custom_metrics.ConfusionMatrix` (see the section below on how test IDs work in ValidMind and why this format is important)\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:527]:
```suggestion
    "- The `@vm.test` decorator is doing the work of creating a wrapper around the function that will allow it to be run by the ValidMind framework. It also registers the test so it can be found by the ID `my_custom_tests.ConfusionMatrix` (see the section below on how test IDs work in ValidMind and why this format is important)\n",
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 7]
```diff
@@ -0,0 +1,675 @@
...
+   "metadata": {},
+   "source": [
+    "### 2.5 Custom test using VMDataset and VMModel as inputs\n",
+    "\n",
+    "We will now create a `@vm.metric` wrapper that will allow you to create a reusable metric. Note the following changes in the code below:\n",
+    "\n",
+    "- The function `confusion_matrix` takes two arguments `dataset` and `model`. This is a `VMDataset` and `VMModel` object respectively.\n",
+    "  - `VMDataset` objects allow you to access the dataset's true (target) values by accessing the `.y` attribute.\n",
+    "  - `VMDataset` objects allow you to access the predictions for a given model by accessing the `.y_pred()` method.\n",
+    "- The function docstring provides a description of what the metric does. This will be displayed along with the result in this notebook as well as in the ValidMind platform.\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:524]:
```suggestion
    "- The function docstring provides a description of what the test does. This will be displayed along with the result in this notebook as well as in the ValidMind platform.\n",
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 8]
```diff
@@ -0,0 +1,675 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn import metrics\n",
+    "import matplotlib.pyplot as plt\n",
+    "@vm.metric(\"my_custom_metrics.ConfusionMatrix\")\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:540]:
```suggestion
    "@vm.test(\"my_custom_metrics.ConfusionMatrix\")\n",
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 9]
```diff
@@ -0,0 +1,675 @@
...
+    "\n",
+    "**Documentation template**: Functions as a test suite and lays out the structure of model documentation, segmented into various sections and sub-sections. Documentation templates define the structure of your model documentation, specifying the tests that should be run, and how the results should be displayed.\n",
+    "\n",
+    "**Tests**: A function contained in the ValidMind Developer Framework, designed to run a specific quantitative test on the dataset or model. Tests are the building blocks of ValidMind, used to evaluate and document models and datasets, and can be run individually or as part of a suite defined by your model documentation template.\n",
+    "\n",
+    "**Metrics**: A subset of tests that do not have thresholds. In the context of this notebook, metrics and tests can be thought of as interchangeable concepts.\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:61]:
```suggestion
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 10]
```diff
@@ -0,0 +1,675 @@
...
+    "\n",
+    "**Custom metrics**: Custom metrics are functions that you define to evaluate your model or dataset. These functions can be registered with ValidMind to be used in the platform.\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:63]:
```suggestion
    "**Custom tests**: Custom tests are functions that you define to evaluate your model or dataset. These functions can be registered with ValidMind to be used in the platform.\n",
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 11]
```diff
@@ -0,0 +1,675 @@
...
+    "\n",
+    "**Inputs**: Objects to be evaluated and documented in the ValidMind framework. They can be any of the following:\n",
+    "\n",
+    "- **model**: A single model that has been initialized in ValidMind with [`vm.init_model()`](https://docs.validmind.ai/validmind/validmind.html#init_model).\n",
+    "- **dataset**: Single dataset that has been initialized in ValidMind with [`vm.init_dataset()`](https://docs.validmind.ai/validmind/validmind.html#init_dataset).\n",
+    "- **models**: A list of ValidMind models - usually this is used when you want to compare multiple models in your custom metric.\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:69]:
```suggestion
    "- **models**: A list of ValidMind models - usually this is used when you want to compare multiple models in your custom test.\n",
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 12]
```diff
@@ -0,0 +1,675 @@
...
+    "- **datasets**: A list of ValidMind datasets - usually this is used when you want to compare multiple datasets in your custom metric. See this [example](https://docs.validmind.ai/notebooks/how_to/run_tests_that_require_multiple_datasets.html) for more information.\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:70]:
```suggestion
    "- **datasets**: A list of ValidMind datasets - usually this is used when you want to compare multiple datasets in your custom test. See this [example](https://docs.validmind.ai/notebooks/how_to/run_tests_that_require_multiple_datasets.html) for more information.\n",
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 13]
```diff
@@ -0,0 +1,675 @@
...
+    "\n",
+    "**Parameters**: Additional arguments that can be passed when running a ValidMind test, used to pass additional information to a metric, customize its behavior, or provide additional context.\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:72]:
```suggestion
    "**Parameters**: Additional arguments that can be passed when running a ValidMind test, used to pass additional information to a test, customize its behavior, or provide additional context.\n",
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 14]
```diff
@@ -0,0 +1,675 @@
...
+    "\n",
+    "**Outputs**: Custom metrics can return elements like tables or plots. Tables may be a list of dictionaries (each representing a row) or a pandas DataFrame. Plots may be matplotlib or plotly figures.\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:74]:
```suggestion
    "**Outputs**: Custom tests can return elements like tables or plots. Tables may be a list of dictionaries (each representing a row) or a pandas DataFrame. Plots may be matplotlib or plotly figures.\n",
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 15]
```diff
@@ -0,0 +1,675 @@
...
+    "    return pd.DataFrame({\"Classe\":classes, \"Percentage\": percentages})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### How to Run the Test\n",
+    "\n",
+    "Let's run the test using the `run_test` method, which is part of the `validmind.tests` module. Here, we pass the `dataset` through the `inputs`. Similarly, you can pass `datasets`, `model`, or `models` as inputs if your custom test requires them. In this example below, we run the custom test `my_custom_metrics.ClassImbalance` by passing the `dataset` through the `inputs`. \n"
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:316]:
```suggestion
    "Let's run the test using the `run_test` method, which is part of the `validmind.tests` module. Here, we pass the `dataset` through the `inputs`. Similarly, you can pass `datasets`, `model`, or `models` as inputs if your custom test requires them. In this example below, we run the custom test `my_custom_tests.ClassImbalance` by passing the `dataset` through the `inputs`. \n"
```

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb [hunk 16]
```diff
@@ -0,0 +1,675 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from validmind.tests import run_test\n",
+    "result = run_test(\n",
+    "    test_id=\"my_custom_metrics.ClassImbalance\",\n",
```
Comment (johnwalz97) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:327]:
```suggestion
    "    test_id=\"my_custom_tests.ClassImbalance\",\n",
```

**Review by johnwalz97 — APPROVED** [review]
Good stuff 🙌 ... Left a few suggestions to update metric to test anywere it is used since @cachafla just merged his PR that does that everywhere else :)

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb
Comment (cachafla) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:1]: I'll add comments inline with screenshots when possible since that's easier to do with a Jupyter notebook.

- The two images are presented without any context, in the middle of the **Key concepts** section:
image

There should ideally be some text accompanying those images, ideally with a title.

- The text here is unrelated to the actual code snippet, which is just initializing a Pandas DataFrame:

image

- The section **Initialize the ValidMind datasets** starts with: `Now, assume we have identified some tests we want to run...`. The previous step only loaded a Pandas DataFrame.

I'd suggest probably moving the code snippet above into the **Initialize the ValidMind datasets** section and explain to the user what is going to happen, i.e. a sample dataframe will get loaded and we will initialize a VMDataset to show how x, y, z can be done, etc.

- Somewhere in the notebook we should explain that it's important to know how to use VMDatasets since that's how you will understand how a custom test works

- On the section **2.1 VMDataset Object** I would go more into detail on the what I consider additional key properties of a VMDataset that the user should be familiar with, if they want to feel more confident when writing unit tests:
    - raw_dataset
    - columns
    - feature_columns
    - feature_columns_numeric
    - feature_columns_categorical
    - text_column. Here, I would an example custom test that shows this in action
    - df
    - prediction_column
    - probability_column

^ I would probably accommodate the existing custom tests you provided to show how they are used. We can still refer the user [here](https://docs.validmind.ai/validmind/validmind/vm_models.html#VMDataset.__init__) for them to read the entire documentation.

In my opinion it's less important to show a realistic example in favor of showing how these dataset properties/methods work.

- This title **2.2 Custom test** should be more descriptive

- Small typo, it should be "Classes" here: `result = pd.DataFrame({"Classe":classes...`

**Review by cachafla — COMMENTED** [review]
Looking good 👌 I left some comments on the notebook.

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb (hunk 4 above)
Comment (AnilSorathiya) [reply notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:353]: done

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb
Comment (juanmleng) [inline notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:1]: Some comments: 

- Perhaps simplify the title of the notebook to "Introduction to ValidMind Dataset and Model Objects."
- Consider making the introduction more specific to VM datasets and models, by providing definitions of these two objects at this stage. Currently, both the intro for model developers and this notebook, share the same introduction.
- The link to available templates in ”You can [view the available templates](https://docs.validmind.com/guide/swap-documentation-templates.html#view-current-templates) to see what has been defined on the platform.” leads to a section that might need to be rewritten to match the title of the section. This section in the docs discusses the template linked to a specific model documentation as opposed to the list of available templates, which can be accessed via Settings and then Templates. This suggestion has a dependency with the docs team.
- Key Concepts section:
    - The content in Key Concepts is also the same in both notebooks, so perhaps tailor this section to target VM datasets and objects more specifically.
    - When describing “models” and “datasets,” note that these concepts apply to both custom and out-of-the-box tests. Perhaps be more general here.
    - Consider moving the Test Suite concept before the documentation template, as the documentation template mentions the test suite as part of the description of the concept.
    - Perhaps use Test Inputs instead of Inputs.
    - The same applies to Parameters; consider using Test Parameters.
    - The same applies to Outputs; consider Test Outputs. Also, the description is linked to the concept of custom metrics, while outputs should be a test concept in general, whether for out-of-the-box or custom tests.
    - We mention out-of-the-box and custom tests; it might be useful to provide a brief description of these two concepts at the beginning of the notebook. 
    - More generally, and just thinking from a comms perspective, I am wondering if perhaps is worth agreeing on a term when referring to out-of-the-box tests -e.g., built-in tests, etc.

**Review by juanmleng — COMMENTED** [review]
LGTM 👍  I just left some styling comments in the notebook.

File: notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb
Comment (AnilSorathiya) [reply notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:1]: @cachafla  I have update the notebook and moved in the `how_to` folder. Can you please review it again?

Comment (AnilSorathiya) [reply notebooks/tutorials/intro_validmind_dataset_model_objects.ipynb:1]: @juanmleng Can you please review the notebook again? It has been moved in the `how_to` folder now.

**Review by cachafla — APPROVED** [review]
Great 👌
